### PR TITLE
Reduce the number of view-changes that occur when a new config is added

### PR DESCRIPTION
**Pull request by ashamis**
Part of #1852

BFT maintains a list of ordered nodes to ensure view-changes are deterministic. This logic needs to also be maintained during reconfiguration. This change adds an offset into every configuration so that if possible the primary will not change between configurations. A complication arises in determining at what view a replica sees a reconfiguration because if a replica is catching up the reconfiguration could have occurred in a different view to the one that the current primary is sending append entries from. To solve this we look at the last committed view-change, this view-change also serves as a deterministic marker on how the offset is created (for auditing purposes)

### CODE REVIEW

File: src/consensus/aft/impl/view_change_tracker.h [hunk 1]
```diff
@@ -239,7 +240,7 @@ namespace aft
       }
 
       auto& vc = it->second;
-      ccf::ViewChangeConfirmation nv(vc.view);
+      ccf::ViewChangeConfirmation nv(vc.view, from);
```
Comment (achamayou) [inline src/consensus/aft/impl/view_change_tracker.h:243]: Is `from` the node id that issued the view change confirmation? Is that effectively the primary at the time?

**Comment by achamayou** [comment]
I'm having a hard time working out what the test for this should look like, but I think we ought to have one. Perhaps get the primary from a config history, and obtain the same result from an arbitrarily truncated sequence with the correct initial view primary id?

**Review by achamayou — APPROVED** [review]

**Comment by ashamis** [comment]
> I'm having a hard time working out what the test for this should look like, but I think we ought to have one. Perhaps get the primary from a config history, and obtain the same result from an arbitrarily truncated sequence with the correct initial view primary id?

Yes I agree with you on this - this is close to what we get with the recovery tests & and in part the rotation/partition tests

File: src/consensus/aft/impl/view_change_tracker.h (hunk 1 above)
Comment (achamayou) [reply src/consensus/aft/impl/view_change_tracker.h:243]: can we rename that to primary?

**Pull request merged**